src/server/replica.h

```diff
@@ -35,48 +51,36 @@ class Replica {
   // Spawns a fiber that runs until link with master is broken or the replication is stopped.
   // Returns true if initial link with master has been established or
   // false if it has failed.
-  bool Run(ConnectionContext* cntx);
+  bool Start(ConnectionContext* cntx);
 
-  // Thread-safe
-  void Stop();
+  void Stop();  // thread-safe
 
-  const std::string& master_host() const {
-    return master_context_.host;
-  }
+  void Pause(bool pause);
 
-  uint16_t port() {
-    return master_context_.port;
-  }
+ private: /* Main standalone mode functions */
+  // Coordinate state transitions. Spawned by start.
+  void MainReplicationFb();
 
-  struct Info {
-    std::string host;
-    uint16_t port;
-    bool master_link_established;
-    bool sync_in_progress;      // snapshot sync.
-    time_t master_last_io_sec;  // monotonic clock.
-  };
+  std::error_code ConnectSocket();  // Connect to master.
+  std::error_code Greet();          // Send PING and REPLCONF.
 
-  // Threadsafe, fiber blocking.
-  Info GetInfo() const;
-  void Pause(bool pause);
+  std::error_code InitiatePSync();     // Redis full sync.
+  std::error_code InitiateDflySync();  // Dragonfly full sync.
 
- private:
-  // Used to initialize df replication flow.
-  Replica(const MasterContext& context, uint32_t flow_id, Service* service);
+  std::error_code ConsumeRedisStream();  // Redis stable state.
+  std::error_code ConsumeDflyStream();   // Dragonfly stable state.
 
-  // The flow is : R_ENABLED -> R_TCP_CONNECTED -> (R_SYNCING) -> R_SYNC_OK.
-  // SYNCING means that the initial ack succeeded. It may be optional if we can still load from
-  // the journal offset.
-  enum State : unsigned {
-    R_ENABLED = 1,  // Replication mode is enabled. Serves for signaling shutdown.
-    R_TCP_CONNECTED = 2,
-    R_GREETED = 4,
-    R_SYNCING = 8,
-    R_SYNC_OK = 0x10,
-  };
+ private: /* Main dlfly flow mode functions */
+  // Initialize as single dfly flow.
+  Replica(const MasterContext& context, uint32_t dfly_flow_id, Service* service);
+
+  // Start replica initialized as dfly flow.
+  std::error_code StartAsDflyFlow();
```

Comment:
Instead of supporting multiple modes and using only one function, I would even suggest splitting up the replica class into three:
- A common replica base (with all state and utils)
- A main replica (with the initiate and consume functions)
- A flow replica (with a separate start)

Reply:
if you think it makes sense and you know how to model it cleanly - do it.